<!--
 * @name: 文件
 * @Author: Haojin Sun
 * @Date: 2020-04-01 13:29:01
 * @LastEditors: Haojin Sun
 * @LastEditTime: 2020-04-01 16:18:10
 -->
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>

<body>
    <script>
        let MyPromise = (function () {
            function resolve(res) {
                this['[[PromiseStatus]]'] = 'fulfilled'
                this['[[PromiseValue]]'] = res
            }

            function reject(res) {
                this['[[PromiseStatus]]'] = 'rejected'
                this['[[PromiseValue]]'] = res
            }
            return function (fn) {
                this['[[PromiseStatus]]'] = 'pending'
                this['[[PromiseValue]]'] = undefined
                fn(resolve.bind(this), reject.bind(this))
            }
        })()

        MyPromise.prototype.then = function (fn) {
            // 只有在成功状态才会调用
            if (this['[[PromiseStatus]]'] === 'fulfilled') {
                this['[[PromiseValue]]'] = fn(this['[[PromiseValue]]'])
            }
            // 返回自身 进行链式调用
            return this
        }
        MyPromise.prototype.catch = function (fn) {
            // 只有在失败状态才会调用
            if (this['[[PromiseStatus]]'] === 'rejected') {
                this['[[PromiseValue]]'] = fn(this['[[PromiseValue]]'])
            }
            // 返回自身 进行链式调用
            return this
        }

        let a = new MyPromise(function (resolve, reject) {
            setTimeout(function () {
                resolve(1)
            }, 1000)
        }).then(function (res) {
            console.log(res)
            return 2
        }).catch(function (res) {
            console.log(res)
            return 3
        })
        console.log(a)
        console.log(new Promise(
            function (resolve, reject) {
                reject(1)
            }
        )
            .then(function (ok) {
                console.log(ok)
            }, function (err) {
                console.log('出错了')
                console.log(err)
                throw new Error('出错啦')
            })
            .then(function(res){
                console.log('我是对的')
                console.log(res)
            })
            .catch(function(res){
                console.log('我是错的1')
                console.log(res)
                throw new Error('出错啦') 
            })
            .catch(function(res){
                console.log('我是错的2')
                console.log(res)
                return res
            })
            .then(function(res){
                console.log('我变成对的拉')
                throw new Error('出错啦') 
            })
            .catch(function(res){
                console.log('我又变成错的拉')
                console.log(res)
                return res
            })
        )
    </script>
</body>

</html>